hi-ml-cpath/testhisto/testhisto/utils/test_callbacks.py

```diff
@@ -93,108 +108,130 @@ def test_loss_analysis_epochs_interval(epochs_interval: int) -> None:
     current_epoch = 5
     assert callback.should_cache_loss_values(current_epoch)
 
+    current_epoch = max_epochs  # no loss caching for extra validation epoch
+    assert callback.should_cache_loss_values(current_epoch) is False
 
-def test_on_train_batch_start(tmp_path: Path, mock_panda_tiles_root_dir: Path) -> None:
+
+def test_on_train_and_val_batch_end(tmp_path: Path, mock_panda_tiles_root_dir: Path) -> None:
     batch_size = 2
-    container = MockDeepSMILETilesPanda(tmp_path=mock_panda_tiles_root_dir)
-    container.batch_size = batch_size
+    container = MockDeepSMILETilesPanda(tmp_path=mock_panda_tiles_root_dir, analyse_loss=True, batch_size=batch_size)
     container.setup()
     container.create_lightning_module_and_store()
 
     current_epoch = 5
     trainer = MagicMock(current_epoch=current_epoch)
-    batch = next(iter(container.data_module.train_dataloader()))
 
     callback = LossAnalysisCallback(outputs_folder=tmp_path)
-    _assert_loss_cache_contains_n_elements(callback.loss_cache, 0)
-
-    callback.on_train_batch_start(trainer, container.model, batch, 0, None)  # type: ignore
-    _assert_loss_cache_contains_n_elements(callback.loss_cache, batch_size)
+    _assert_loss_cache_contains_n_elements(callback.loss_cache[ModelKey.TRAIN], 0)
+    _assert_loss_cache_contains_n_elements(callback.loss_cache[ModelKey.VAL], 0)
 
-    batch = next(iter(container.data_module.train_dataloader()))
-    callback.on_train_batch_start(trainer, container.model, batch, 1, None)  # type: ignore
-    _assert_loss_cache_contains_n_elements(callback.loss_cache, 2 * batch_size)
+    def _call_on_batch_end_hook(on_batch_end_hook: Callable, batch_idx: int) -> None:
+        batch = next(iter(container.data_module.train_dataloader()))
```

Comment:
[Minor] Doesn't this always return the same batch every time? If so, won't the cache have duplicates, and is it a problem at all?

Reply:
Yes, I am aware! the reason why I am not using a new batch each time is because I don't have enough batches with the current fixture [mock_data](https://github.com/microsoft/hi-ml/blob/79ced59ca53b3a7b32074734647ffe310d109441/hi-ml-cpath/testhisto/conftest.py#L81) to stimulate 2 train steps + 2 val steps, I had a StopIteration ERROR but maybe I can just increase the number of slides. (we currently have 10 slides, and only 10% for validation, also the reason why I'm using the same dataloader for both train and val).